### PR TITLE
Interface Analyzer

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

- [x] This PR adds in a static analysis tool, which parses the AST for the selected interfaces and then checks if the appropriate
equality checks are being compared.
- [x] Fixes across prysm for issues found by the static analyzer.

**Which issues(s) does this PR fix?**

N.A

**Other notes for review**

### CODE REVIEW

**Comment by rkapka** [comment]
Why do you add a new check instead of replacing the existing one? After all `IsNIl` does include a nilness check on the receiver.

**Comment by nisdas** [comment]
@rkapka because the interface itself could be nil, so running any receiver methods would create a panic

**Comment by rkapka** [comment]
OK but in that case the `b == nil` part in `return b == nil || b.state == nil` doesn't make sense, right? Because `b` can never be nil if we got to this point.

File: tools/analyzers/interfacechecker/analyzer.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		xIsNil := typeMap[identX].IsNil()
+		yisIface := strings.Contains(typeMap[identY].Type.String(), iface)
+		yIsNil := typeMap[identY].IsNil()
+		// Exit early if neither are of the desired interface
+		if !xIsIface && !yisIface {
+			continue
+		}
+		if xIsIface && yIsNil {
+			pass.Reportf(identX.Pos(), "A nilness check is being performed on an interface"+
+				", this check needs another accompanying check on the underlying object for the interface")
```
Comment (terencechain) [inline tools/analyzers/interfacechecker/analyzer.go:86]: Little confusing. What does "another accompanying check" mean? Better to be specific

**Review by terencechain — APPROVED** [review]
Nice job!

**Pull request merged**